src/OpenTelemetry/Trace/ExceptionProcessor.cs

```diff
@@ -0,0 +1,80 @@
+// <copyright file="ExceptionProcessor.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Diagnostics;
+using System.Linq.Expressions;
+using System.Reflection;
+using System.Runtime.InteropServices;
+
+namespace OpenTelemetry.Trace
+{
+    internal class ExceptionProcessor : BaseProcessor<Activity>
+    {
+        private const string ExceptionPointersKey = "otel.exception_pointers";
+
+        private readonly Func<IntPtr> fnGetExceptionPointers;
+
+        public ExceptionProcessor()
+        {
+            try
+            {
+                var flags = BindingFlags.Static | BindingFlags.Public;
+                var method = typeof(Marshal).GetMethod("GetExceptionPointers", flags, null, new Type[] { }, null);
+                var lambda = Expression.Lambda<Func<IntPtr>>(Expression.Call(method));
+                this.fnGetExceptionPointers = lambda.Compile();
+            }
+            catch (Exception ex)
+            {
+                throw new NotSupportedException("System.Runtime.InteropServices.Marshal.GetExceptionPointers is not supported.", ex);
+            }
+        }
+
+        /// <inheritdoc />
+        public override void OnStart(Activity activity)
+        {
+            var pointers = this.fnGetExceptionPointers();
+
+            if (pointers != IntPtr.Zero)
+            {
+                activity.SetTag(ExceptionPointersKey, pointers);
+            }
+        }
+
+        /// <inheritdoc />
+        public override void OnEnd(Activity activity)
+        {
+            var pointers = this.fnGetExceptionPointers();
+
+            if (pointers == IntPtr.Zero)
+            {
+                return;
+            }
+
+            var snapshot = activity.GetTagValue(ExceptionPointersKey) as IntPtr?;
```

Comment:
This is going to be expensive. [ActivityTagsCollection](https://github.com/dotnet/runtime/blob/3bb4ce433df67749295e8e77defde7aef5e75c6d/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityTagsCollection.cs#L24) is really just a list. So in the case of no exception in OnStart, which is the more common case, we'll have to loop through all tags here just to exit out without doing anything.

Reply:
Yep. I intended to put it after the `if (pointers == IntPtr.Zero)` check, so we only pay for the cost when there is an exception on the thread.

Reply:
An alternative solution would be use AsyncLocal + ref counting, similar like the [`SuppressInstrumentation` logic](#1202) but it comes with new problems.

Reply:
Ah OK yes it should be fine after the `pointers` check, we'll only pay the cost for the uncommon case.